src/cli/cli_otns.cpp

```diff
@@ -0,0 +1,340 @@
+        mCollectionTreeTimer.Start(Random::NonCrypto::GetUint32InRange(0, mCollectionTreeIntervalMS));
+        ExitNow(error = OT_ERROR_NONE);
+    }
+    else if (strcmp(argv[0], "stop") == 0)
+    {
+        mInterpreter.mServer->OutputFormat("stopping\r\n");
+        mCollectionTreeTimer.Stop();
+        ExitNow(error = OT_ERROR_NONE);
+    }
+
+exit:
+    return error;
+}
+
+otError OTNS::Process(int argc, char *argv[])
+{
+    otError error = OT_ERROR_PARSE;
+
+    if (argc < 1)
+    {
+        ProcessHelp(0, NULL);
+        error = OT_ERROR_INVALID_ARGS;
+    }
+    else
+    {
+        for (size_t i = 0; i < OT_ARRAY_LENGTH(sCommands); i++)
+        {
+            if (strcmp(argv[0], sCommands[i].mName) == 0)
+            {
+                error = (this->*sCommands[i].mCommand)(argc - 1, argv + 1);
+                break;
+            }
+        }
+    }
+    return error;
+}
+
+OT_TOOL_PACKED_BEGIN
+struct sense_data_t
+{
+    uint32_t sensor_id;
+    uint32_t milli_time;
+    uint32_t sense_val;
+} OT_TOOL_PACKED_END;
+
+void OTNS::HandleCollectionTreeTimer(Timer &aTimer)
+{
+    //    otLogCritCore("Sense & collect ...");
+    TimerMilliContext &timer        = static_cast<TimerMilliContext &>(aTimer);
+    OTNS &             otns         = *static_cast<OTNS *>(timer.GetContext());
+    Interpreter &      mInterpreter = otns.mInterpreter;
+
+    // reset the timer
+    uint32_t interval = otns.mCollectionTreeIntervalMS;
+    interval -= interval / 10;
+    interval += Random::NonCrypto::GetUint32InRange(0, interval / 10 * 2);
+    timer.Start(interval);
+
+    // now, send the coap message
+
+    otError       error   = OT_ERROR_NONE;
+    otMessage *   message = NULL;
+    otMessageInfo messageInfo;
+
+    // Default parameters
+    const otCoapType coapType = OT_COAP_TYPE_CONFIRMABLE;
+    otCoapCode       coapCode = OT_COAP_CODE_POST;
+    otIp6Address     coapDestinationIp;
+
+    // Destination IPv6 address
+    //    const otMeshLocalPrefix *mlprefix = otThreadGetMeshLocalPrefix(mInterpreter.mInstance);
+    SuccessOrExit(error = otThreadGetLeaderAloc(mInterpreter.mInstance, &coapDestinationIp));
+    //    SuccessOrExit(error = otIp6AddressFromString("fd00:db8::ff:fe00:fc00", &coapDestinationIp));
+
+    message = otCoapNewMessage(mInterpreter.mInstance, NULL);
+    VerifyOrExit(message != NULL, error = OT_ERROR_NO_BUFS);
+
+    otCoapMessageInit(message, coapType, coapCode);
+    otCoapMessageGenerateToken(message, ot::Coap::Message::kDefaultTokenLength);
+    SuccessOrExit(error = otCoapMessageAppendUriPathOptions(message, "otns/ct"));
+
+    SuccessOrExit(error = otCoapMessageSetPayloadMarker(message));
+
+    sense_data_t sense_data;
+    assert(otns.mCollectionTreeSensorId > 0);
+    sense_data.sensor_id  = otns.mCollectionTreeSensorId;
+    sense_data.milli_time = otPlatAlarmMilliGetNow();
+    sense_data.sense_val  = 0xDDCCBBAA;
+
+    SuccessOrExit(error = otMessageAppend(message, &sense_data, sizeof(sense_data_t)));
```

Comment:
Is it possible that the receiver of this message could be another host? If true, I think there will be endianness problem. If false, I still don't think it is a good idea to send data in platform-dependent form through network.

Reply:
Why mark this resolved?

Reply:
Sorry, only noticed 0xDDCCBBAA.